core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -438,6 +443,11 @@ private[spark] class TaskSetManager(
         blacklist.isExecutorBlacklistedForTaskSet(execId)
     }
     if (!isZombie && !offerBlacklisted) {
+      if (!sortedPendingTasks.get()) {
```

Comment:
I think you need `if (sortedPendingTasks.compareAndSet(false, true))` or else the point of AtomicBoolean is kind of lost